core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -705,74 +727,77 @@ public SourceDistributionTaskSource(
         }
 
         @Override
-        public List<TaskDescriptor> getMoreTasks()
+        public ListenableFuture<List<TaskDescriptor>> getMoreTasks()
         {
             if (finished || closed) {
-                return ImmutableList.of();
+                return immediateFuture(ImmutableList.of());
             }
 
-            List<TaskDescriptor> result = new ArrayList<>();
-
-            boolean splitSourceFinished = false;
-            while (result.isEmpty()) {
-                ListenableFuture<SplitBatch> splitBatchFuture = splitSource.getNextBatch(NOT_PARTITIONED, Lifespan.taskWide(), splitBatchSize);
+            List<TaskDescriptor> result = getAllReadyTasks();
```

Comment:
I'm not sure if we really need this as a special case. I would just let it go the normal path and inline the `getAllReadyTasks` to make things easier to reason about